jupyterhub/templates/_helpers.tpl

```diff
@@ -173,10 +173,39 @@ component: {{ include "jupyterhub.componentLabel" . }}
 {{ include "jupyterhub.matchLabels" $_ | replace ": " "=" | replace "\n" "," | quote }}
 {{- end }}
 
+
 {{- /*
-  singleuser.imagePullSecret:
-    allows creating a base64 encoded docker registry json blob
+  jupyterhub.dockerconfigjson:
+    Creates a base64 encoded docker registry json blob for use in a image pull
+    secret, just like the `kubectl create secret docker-registry` command does
+    for the generated secrets data.dockerconfigjson field. The output is
+    verified to be exactly the same even if you have a password spanning
+    multiple lines as you may need to use a private GCR registry.
+
+    - https://kubernetes.io/docs/concepts/containers/images/#specifying-imagepullsecrets-on-a-pod
 */}}
-{{- define "singleuser.imagePullSecret" }}
-{{- printf "{\"auths\": {\"%s\": {\"auth\": \"%s\"}}}" .Values.singleuser.imagePullSecret.registry (printf "%s:%s" .Values.singleuser.imagePullSecret.username .Values.singleuser.imagePullSecret.password | b64enc) | b64enc }}
+{{- define "jupyterhub.dockerconfigjson" -}}
+{{ include "jupyterhub.dockerconfigjson.yaml" . | b64enc }}
+{{- end }}
+
+{{- define "jupyterhub.dockerconfigjson.yaml" -}}
+{{- with . -}}
+{
+  {{- /* */ -}}
+  "auths":{
+    {{- .registry | default "https://index.docker.io/v1/" | quote }}:{
+      {{- /* */ -}}
```

Comment:
What are all these empty comments doing?

Reply:
I wanted to recreate the exact output of when using kubectl to create the secret in order to ensure things works. But that output was very compact, without any whitespace or newlines. The comments allow me to whitespace chomp the added spaces and newlines that makes this somewhat readable at all. Without them it became a blob of text i could not read.

We could make the output contain the extra whitepace though, my guess is that it would still work, im not 100% on it though so i figured it was best not to deviate at all.

Reply:
sunmary: they make a blob without whitespace somewhat readable. if removed we deviate from generated output of kubectl doing the same thing, which may be fine but i dont know for sure

Reply:
Deal! Thanks for the decisive input!